File: Leetcode/209.长度最小的子数组.cpp
/*
 * @lc app=leetcode.cn id=209 lang=cpp
 *
 * [209] 长度最小的子数组
 */
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;
/*2021-08-16-15-45*/
// ? 思路解析：
/*
 * 本来是在 二分tag 下看到的，结果前缀和加二分不会写
 *
 ! yxc 双指针写法，很厉害，i j 具有单调性的问题就可以使用双指针来解决 o(n)
 !

static_cast 用于良性转换，一般不会导致意外发生，风险很低。
c++的四种转换，http://c.biancheng.net/cpp/biancheng/view/3297.html
 */

// @lc code=start
class Solution {
   public:
    int minSubArrayLen(int target, vector<int>& nums) {
        int ans = INT_MAX;
        vector<int> s(nums.size() + 1, 0);
        for (int i = 1; i <= nums.size(); i++) s[i] = s[i - 1] + nums[i - 1];
        for (int i = 0; i < s.size(); i++) {
            int l = i, r = s.size() - 1;
            while (l < r) {
                int mid = (l + r) >> 1;
                if (s[mid] - s[i] >= target)
                    r = mid;
                else
                    l = mid + 1;
            }
            if (s[l] - s[i] >= target) ans = min(ans, l - i);
        }
        return ans == INT_MAX? 0 : ans;
    }
};
// @lc code=end

class Solution_双指针 {
   public:  // O(n)
    int minSubArrayLen(int target, vector<int>& nums) {
        int asn = INT_MAX;
        //  i j 为双指针， sum 是 i 到 j 的和
        for (int i = 0, j = 0, sum = 0; i < nums.size(); i++) {
            sum += nums[i];
            while (sum - nums[j] >= target) sum -= nums[j++];
            //! 如果 while 循环可以执行的话
            //! 说明 新加入的 i 可以至少电梯一个去掉的 j
            // j 向后移动，找到满足 ij 满足条件的最右侧 j 的值
            if (sum >= target) asn = min(asn, i - j + 1);
        }
        if (asn == INT_MAX) asn = 0;
        return asn;
    }
};

class Solution_前缀和加二分查找 {
   public:  // O(n * log_n)
    int minSubArrayLen(int target, vector<int>& nums) {
        int ans = INT_MAX;
        vector<int> s(nums.size() + 1, 0);
        vector<int> a(nums.size() + 1, 0);
        for (int i = 1; i <= nums.size(); i++) a[i] = nums[i - 1];
        for (int i = 1; i <= nums.size(); i++) s[i] = s[i - 1] + a[i];
        // // for (auto i : s) cout << i << ' ';
        // // puts("");
        // s[i] 代表前缀和数组， a[i] 代表区间元素

        for (int i = 0; i < s.size(); i++) {
            // ! 注意 i 要从 0 开始，因为还有当区间和刚好等于目标值的情况
            int l = i, r = s.size() - 1;
            // 使用二分优化找到区间起始位置时，再找区间结束位置的步骤
            while (l < r) {
                int mid = (l + r) >> 1;
                if (s[mid] - s[i] >= target)
                    r = mid;
                else
                    l = mid + 1;
            }
            // // cout << " i = " << i << " l = " << l << endl;
            if (s[l] - s[i] >= target) {
                // // cout << s[l] - s[i] << ">> 区间长度 " << l - i << endl;
                ans = min(ans, l - i);
                // !  l - i + 1 是 s 数组的长度，不加 1 是 a 数组的长度
            }
        }
        return ans == INT_MAX ? 0 : ans;
    }
};


class Solution_暴力解法 {
   public:  // O(n^2)
    int minSubArrayLen(int s, vector<int>& nums) {
        int n = nums.size();
        if (n == 0) return 0;

        int ans = INT_MAX;
        for (int i = 0; i < n; i++) {
            int sum = 0;
            for (int j = i; j < n; j++) {
                sum += nums[j];
                if (sum >= s) {
                    ans = min(ans, j - i + 1);
                    break;
                }
            }
        }
        return ans == INT_MAX ? 0 : ans;
    }
};

File: Leetcode/209.长度最小的子数组.go
/*
 * @lc app=leetcode.cn id=209 lang=golang
 *
 * [209] 长度最小的子数组
 *
 * https://leetcode.cn/problems/minimum-size-subarray-sum/description/
 *
 * algorithms
 * Medium (46.75%)
 * Likes:    727
 * Dislikes: 0
 * Total Accepted:    172.8K
 * Total Submissions: 369.5K
 * Testcase Example:  '7\n[2,3,1,2,4,3]'
 *
 * 给定一个含有 n 个正整数的数组和一个正整数 target 。
 *
 * 找出该数组中满足其和 ≥ target 的长度最小的 连续子数组 [numsl, numsl+1, ..., numsr-1, numsr]
 * ，并返回其长度。如果不,存在符合条件的子数组，返回 0 。
 *
 *
 *
 * 示例 1：
 *
 *
 * 输入：target = 7, nums = [2,3,1,2,4,3]
 * 输出：2
 * 解释：子数组 [4,3] 是该条件下的长度最小的子数组。
 *
 *
 * 示例 2：
 *
 *
 * 输入：target = 4, nums = [1,4,4]
 * 输出：1
 *
 *
 * 示例 3：
 *
 *
 * 输入：target = 11, nums = [1,1,1,1,1,1,1,1]
 * 输出：0
 *
 *
 *
 *
 * 提示：
 *
 *
 * 1
 * 1
 * 1
 *
 *
 *
 *
 * 进阶：
 *
 *
 * 如果你已经实现 O(n) 时间复杂度的解法, 请尝试设计一个 O(n log(n)) 时间复杂度的解法。
 *
 *
 */
package leetcode

/* --- 2022-01-20-21-21 --- */

// https://github.com/fengwei2002/Algorithm

// solution link:
// https://leetcode.cn/problems/minimum-size-subarray-sum/solution/lc209-fengwei2002-by-kycu-mfnn/

// @lc code=start
func minSubArrayLen(target int, nums []int) int {
	// 找出数组中和大于等于 target 的长度最小的子数组
	ans := (int)(1e9)
	for i, j, sum := 0, 0, 0; i < len(nums); i++ {
		sum += nums[i]
		for sum-nums[j] >= target {
			sum -= nums[j]
			j++
		}
		if sum >= target {
			ans = min(ans, i-j+1)
		}
	}
	if ans == (int)(1e9) {
		ans = 0
	}
	return ans
}

func min(a, b int) int {
	if a > b {
		return b
	}
	return a
}

// @lc code=end
